src/db/models/attachmentPrototype.ts

```diff
@@ -37,15 +37,35 @@ const FIELDS = t.array(
 export const ATTACHMENT_PROTOTYPE_VALUE = t.intersection([
   // Required Fields
   t.type({
-    hasMeasures: t.number,
+    /**
+     * Some prototypes incorrectly quote numbers, which we need to account for
+     *
+     * TODO: clean prototype data and make type stricter
+     */
+    hasMeasures: t.union([
+      t.literal(0),
+      t.literal(1),
+      t.literal('0'),
+      t.literal('1'),
+    ]),
```

Comment:
This is actually not representing what data **can** be stored. Maybe it is true for the data that **is stored already**, I haven't checked (btw, do you know any efficient way to query JSON columns in Postgres to retrieve patterns), and maybe possibly true for the data we **want** stored (except for string numbers).

On the other hand, RPM UI allows for any value as far as I can see. I just tried putting number 2 and letter 'a' and both got stored nicely. With validation, we will prevent such cases, but it should be handled on the UI side as well.

We just need to make sure that there are no outliers for this field, then write admin command and clean prototype data.

But, then again, this should actually be boolean. Even if we allowed for any number and only treat value 1 as true and all other values as false is wrong and has bitten us already.

So, I think the right approach would be to tackle this in a separate ticket. We should write admin command that does the "migration" (person who does this should also inspect the data to see whether there are any strange pieces of data), and moves all falsy values to `false` (not `'false'`!) and truthy values to `true` (not `'true'`). Other strategy might also be applied, really depends on how much invalid values we have for `hasMeasures`.

Then, we only need to update this type definition here and everything should be fine. RPM already uses a library (I think `jsoneditor`) that will display a nice check box for us.
The only problem I see with this approach is internationalization.

Reply:
Agreed on all counts. Though i don't feel the internationalization issue is a problem, if a user is capable of writing a json prototype, and know json syntax, then they know about `true` and `false`.

And right, as you say any data can be stored right now, but at the very least this covers the majority of cases for current active data that's stored in RPM, which means that it can be used, and if any new invalid data is stored, then it will fail gracefully by reporting the invalid data and 500ing.

We definitely want the validation to occur upon saving / updating these attachments. IMO the validation can still be defined and occur server-side, as long as an error is propigated up to the frontend and displayed to the user somehow. So we want to update the 

Right now the v4 library doesn't do any runtime validation upon save (instead relying on the type safety of the caller), but if we wanted to start using this model from javascript endpoints (or any javascript data that doesn't go via a typescript `unknown` type forcing us to validate before sending it to the model), then we may also want to add runtime type validation to `create` and `update` in `raw-model.ts`.

Created an issue here to track: https://humanitarian.atlassian.net/browse/HPC-8258

Reply:
Thank you for creating the ticket.

What do you think about doing data validation before inserting/updating in `create`/`update`? That would require validation of user data. What we do now is insert/update first and call `validateAndFilter` on instance created/updated.

While that is good for `find`, we maybe want to have runtime type validation for user data supplied. I am also suspicious of this, because we may need to update a lot of the code as it will reveal lots of bugs, since many json fields provide more data than needed. Also, there are many numbers stored as strings.

In `validateAndFilter`, `t.exact` does the filtering, but I think codec check would fail if we do this before actual save. What is your thinking on this?

> So we want to update the

This sentence was left incomplete :)